src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt

```diff
@@ -33,19 +38,36 @@ class RsImplIndex : AbstractStubIndex<TyFingerprint, RsImplItem>() {
             val fingerprint = TyFingerprint.create(target)
                 ?: return emptyList()
 
-            return getElements(KEY, fingerprint, project, GlobalSearchScope.allScope(project))
+            val impls = getElements(KEY, fingerprint, project, GlobalSearchScope.allScope(project))
                 .filter { impl ->
                     val ty = impl.typeReference?.type
                     // Addition class check is a temporal solution to filter impls for type parameter
                     // with the same name
                     // struct S; impl<S: Tr1> Tr2 for S {}
                     ty != null && ty.javaClass == target.javaClass
                 }
-            }
+            val freeImpls = getElements(KEY, TyFingerprint.TYPE_PARAMETER_FINGERPRINT, project, GlobalSearchScope.allScope(project))
+            return impls + freeImpls
+        }
 
         fun index(stub: RsImplItemStub, sink: IndexSink) {
-            val type = stub.psi.typeReference ?: return
-            val key = TyFingerprint.create(type)
+            val impl = stub.psi
+            val typeRef = impl.typeReference ?: return
+            val type = typeRef.typeElement
+
+            val key = if (type is RsBaseType) {
+                val typeParam = impl.typeParameters.find { it.name == type.name }
+                // At this moment we support only free trait impls without bounds
+                // to avoid stack overflow while resolving
+                if (typeParam != null && typeParam.bounds.isEmpty()) {
+                    TyFingerprint.TYPE_PARAMETER_FINGERPRINT
```

Comment:
If type params is not empty we should just ignore the impl (exclude from index)
Also, what about whew clauses?

Reply:
`RsTypeParameter.bounds` is extension property (see [here](https://github.com/intellij-rust/intellij-rust/blob/4c7c1e331e49b84aa262fcc7471cef72605a3920/src/main/kotlin/org/rust/lang/core/psi/ext/RsTypeParameter.kt#L17-L25])) which collect all bounds including where clause

Reply:
Oh, OK